3단계 - 지하철 구간 관리

giwankim opened the pull request:
2단계 코드 리뷰 반영하고 구간 관리 기능을 구현했습니다.

약간 생각보다 시간이 들었는데 😅 또 좋은 코드리뷰 부탁드리겠습니다!  🙇

giwankim committed:
refactor(test): 사용 안 하는 코드를 제거한다.

giwankim committed:
refactor(station): 함수명을 일관되게 변경한다.

`StationReader#read()` -> `StationReader#readById()`

giwankim committed:
fix(line): 노선을 못 찾았을 경우 예외 메시지에 ID를 추가한다.

giwankim committed:
refactor(line): inline function

giwankim committed:
refactor(line): `StationResponse` DTO 클래스를 재활용한다.

giwankim committed:
feat(support): 예외 처리와 관련된 클래스들을 추가한다.

giwankim committed:
refactor(line): `ApiException`을 상속한 커스텀 예외 클래스를 던진다.

giwankim committed:
refactor(station): `ApiException`을 상속한 커스텀 예외 클래스를 던진다.

giwankim committed:
refactor(line): DELETE /lines/{id} 인수 테스트를 HttpHeader 값으로 확인한다.

수정 전에는 전체 노선 목록을 조회해 와서 확인했다.

giwankim committed:
feat(section): 구간 추가를 위한 인수 테스트 뼈대 추가

미션 특성상 예외 처리 확인을 단위 테스트 대신 인수 테스트로 처리할 수 있기에 우선 빈 테스트로 추가한다.

giwankim committed:
feat(section): 지하철 노선 엔티티를 추가한다.

giwankim committed:
refactor(line): 노선 구간을 사용하도록 지하철 노선 생성 로직을 리팩토링한다.

giwankim committed:
refactor(line): 노선의 구간 목록을 일급 콜렉션으로 wrapping한다.

giwankim committed:
fix(line): 변경된 테이블 구조에 따라 구간 관리 인수테스트 fixture를 수정했다.

giwankim committed:
feat(line): 노선 구간 추가 기능을 구현했다.

giwankim committed:
reafactor(test): 구간 등록 인수테스트 step 추출한다.

giwankim committed:
test(section): 구간 등록 인수테스트 연장 불가 에러 케이스 작성했다.

giwankim committed:
feat(section): 구간 등록 시 validation을 하는 컴포넌트를 추가했다.

giwankim committed:
test(section): cycle이 생겨서 구간 등록이 실패하는 경우의 인수 테스트를 추가했다.

giwankim committed:
feat(section): 노선에 구간 추가 시 cycle이 생기면 예외가 던져지게 구현했다.

giwankim committed:
feat(section): 구간 제거 기능을 구현했다.

giwankim committed:
refactor(section): 구간 제거 인수 테스트를 단계를 리팩토링했다.

giwankim committed:
feat(section): 종점 역이 아닌 구간 제거 시 에러를 응답하도록 구현했다.

giwankim committed:
test(section): 구간이 하나만 있는 경우 구간 제거 시 에러가 발생한다는 인수 테스트를 작성했다.

giwankim committed:
feat(section): 구간이 하나만 있는 경우 구간 제거 시 예외를 던지는 validation 로직을 구현했다.

seondongpyo commented on src/main/java/subway/line/AppendLineSectionValidator.java:
```diff
@@ -0,0 +1,32 @@
+
+@Component
+@RequiredArgsConstructor
+public class AppendLineSectionValidator {
```
`-Validator`에서 처리하는 유효성 검증 로직들은 도메인 객체 내부로 옮겨보면 어떨까요?

seondongpyo commented on src/main/java/subway/line/LineSections.java:
```diff
@@ -0,0 +1,65 @@
+@Embeddable
+@Getter
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+public class LineSections {
```
일급 컬렉션 활용 👍

seondongpyo commented on src/main/java/subway/line/LineSections.java:
```diff
@@ -0,0 +1,65 @@
+    return Collections.unmodifiableList(stations);
+  }
+
+  public void removeStation(Station station) {
```
실제로 사용되지 않는 매개변수라면 제거해보면 어떨까요?
```suggestion
  public void removeStation() {
```

seondongpyo commented on src/main/java/subway/support/error/ErrorType.java:
```diff
@@ -0,0 +1,17 @@
+public enum ErrorType {
+  DEFAULT(HttpStatus.INTERNAL_SERVER_ERROR, ErrorCode.E500, "서버 내부 에러가 발생했습니다."),
+  BAD_REQUEST(HttpStatus.BAD_REQUEST, ErrorCode.E400, "잘못된 요청입니다."),
+  NOT_FOUND(HttpStatus.NOT_FOUND, ErrorCode.E404, "존재하지 않는 정보입니다.");
```
- `HttpStatus`의 응답 코드와 `ErrorCode`로 지정하신 에러 코드가 거의 동일해보이는데, 
`ErrorCode`를 따로 정의하신 이유가 있으실까요?
- 메시지가 조금 더 구체적이었으면 좋겠습니다. 예를 들면, 존재하지 않는 식별자로 지하철 노선을 조회할 경우 `LineNotFoundException` 예외가 던져지는데 어떤 식별자에 의한 것인지를 예외 메시지에 포함하면 디버깅에 도움이 되지 않을까 생각합니다.

seondongpyo commented on src/test/java/subway/line/AppendLineSectionAcceptanceTest.java:
```diff
@@ -0,0 +1,84 @@
+
+    ExtractableResponse<Response> response = 노선_구간_등록_요청(lineId, request);
+
+    노선_구간_등록됨(response, lineId, downStationId);
```
테스트 메서드만 보자면 노선과 상행역, 하행역이 id로만 선언되어 있기 때문에, 어떤 지하철 노선에 어떤 구간을 등록하는지를 쉽게 이해하기 어려울 것 같아요. 인수 테스트의 가독성을 높여볼 수 있는 방법이 없을까요?

seondongpyo commented on src/test/java/subway/line/AppendLineSectionAcceptanceTest.java:
```diff
@@ -0,0 +1,84 @@
+  }
+
+  /** Given 구간의 하행 역이 이미 해당 노선에 등록되어 있으면 When 구간 등록을 하면 Then 400 Bad Request 에러가 반환된다. */
+  @DisplayName("Cycle이 생길 수 있는 지하철 구간 등록 시 에러가 발생한다.")
```
`Cycle`이란 용어 대신에, 위의 주석에 작성된 표현을 사용하는 게 조금 더 이해하기 쉬울 것 같아요.

seondongpyo commented on src/test/java/subway/line/RemoveLineSectionAcceptanceTest.java:
```diff
@@ -0,0 +1,68 @@
+import subway.support.AcceptanceTest;
+
+@DisplayName("지하철 구간 제거 기능")
+class RemoveLineSectionAcceptanceTest extends AcceptanceTest {
```
다음 경우에 대한 인수 테스트도 추가해보면 어떨까요?
`지하철 노선에 상행 종점역과 하행 종점역만 있는 경우(구간이 1개인 경우) 역을 삭제할 수 없다.`

seondongpyo reviewed (changes requested):
마지막 단계도 잘 구현해주셨네요 👍 
몇 가지 코멘트 드렸는데 확인 후 다시 리뷰 요청을 부탁 드리겠습니다 🔥

giwankim committed:
refactor(section): 노선 구간 등록 validation 로직을 도메인으로 옯긴다.

생각해보니 Validator 컴포넌트가 필요한 경우는 다른 애그리거트의 엔티티가 발리데이션 로직에 필요한 경우에 사용하면 될 것 같다. ![OrderValidator.java](https://github.com/eternity-oop/Woowahan-OO-03-domain-event/blob/master/src/main/java/org/eternity/food/order/domain/OrderValidator.java)

giwankim committed:
refactor(section): 노선 구간 등록 밸리데이션 로직을 도메인으로 옮겼다.

giwankim committed:
refactor(section): 구간 등록 인수 테스트의 가독성을 위해서 도메인 픽스쳐를 활용한다.

giwankim committed:
refactor(section): 노선 관리 기능 인수 테스트에서 도메인 픽스쳐를 활용한다.

giwankim committed:
refactor(section): 구간 제거 기능 인수 테스트를 도메인 픽스쳐를 사용하는 방향으로 리팩토링했다.

giwankim committed:
refactor(section): API 에러 메시지를 구체화했다.

seondongpyo commented on src/main/java/subway/station/StationReader.java:
```diff
@@ -8,9 +8,9 @@
     return stationRepository
         .findById(id)
-        .orElseThrow(() -> new IllegalArgumentException("해당 역이 존재하지 않습니다."));
+        .orElseThrow(() -> new StationNotFoundException("역 #" + id + "이 존재하지 않습니다."));
```
예외 메시지 내용은 예외 클래스 내부로 옮겨보면 어떨까요?
```suggestion
        .orElseThrow(() -> new StationNotFoundException(id));
```

seondongpyo commented on src/main/java/subway/line/LineSections.java:
```diff
@@ -0,0 +1,106 @@
+  }
+
+  private boolean isAppendable(LineSection lineSection) {
+    return getLast().getDownStation().isSame(lineSection.getUpStation());
```
아래의 `isTerminalStation()`을 활용해봐도 좋을 것 같습니다
```suggestion
    return isTerminalStation(lineSection.getUpStation());
```

seondongpyo reviewed (approved):
피드백 잘 반영해주셨네요 👍 
이번 미션은 여기서 마무리하도록 하겠습니다.
다음 미션도 힘내시고 마지막 미션까지 꼭 완주하시길 응원하겠습니다 🔥
